### PR TITLE
feat: expand app collection

### PR DESCRIPTION
**Pull request by kleinfreund**
## Changes

**feat: adds custom toolbar styles**

Adds custom toolbar styles so that toolbar entries are automatically placed and spaced in the right position above the table.

**refactor: moves cell attributes function into script block**

**chore: aid syntax highlighting by wrapping `as string` in template**

**chore: marks `total` as optional prop**

Signed-off-by: Philipp Rudloff <philipp.rudloff@konghq.com>

### CODE REVIEW

File: src/app/application/components/app-collection/AppCollection.vue [hunk 1]
```diff
@@ -54,41 +59,117 @@ type ChangeValue = {
...
+      const hasItem = items.find((item) => item && typeof item === 'object' && 'id' in item && item.id === props.selectedRowId)
+
+      if (hasItem) {
+        currentSelectedRowId.value = props.selectedRowId
+      }
+    } else {
+      // Use the first item as a fallback.
+      const item = items[0]
+
+      if (item && typeof item === 'object' && 'id' in item && typeof item.id === 'string') {
```
Comment (kleinfreund) [inline src/app/application/components/app-collection/AppCollection.vue:113]: **Comment**: Not in love with the necessity for that degree of type-checking this value.

I’d rather type `props.items` as a list of objects that at the very least have an `id` property (so we can standardize on _one_ behavior for selecting an active row). But I didn’t want to add this here because it would increase the scope.

File: src/app/application/components/app-collection/AppCollection.vue [hunk 2]
```diff
@@ -54,41 +59,117 @@ type ChangeValue = {
...
+  if ('id' in row && row.id === currentSelectedRowId.value) {
+    attributes.class = 'is-selected'
+  }
+
+  return attributes
+}
+
+function handleRowClickEvent(_event: any, row: any) {
+  currentSelectedRowId.value = row.id
+  emit('row:click', row)
```
Comment (kleinfreund) [inline src/app/application/components/app-collection/AppCollection.vue:150]: **Comment**: `AppCollection` should, at the very least currently, not contain the _behavior_ attached to row clicks. Specifically, navigating to the detail view on row click currently doesn’t make sense because almost none of our list views behave that way: instead, they select the item to show in the summary.

File: src/app/application/components/app-collection/AppCollection.vue [hunk 3]
```diff
@@ -13,34 +13,39 @@
...
+      #toolbar
+    >
+      <div class="app-collection-toolbar">
+        <slot name="toolbar" />
+      </div>
+    </template>
+
+    <template
+      v-for="key in Object.keys(remainingSlots)"
+      #[`${key}`]="{ row, rowValue }"
```
Comment (johncowen) [inline src/app/application/components/app-collection/AppCollection.vue:33]: Could we just check if `key` equals `toolbar` and if so add the wrapping div? In fact from what I can see the KTable toolbar already has a wrapping `k-table-toolbar`, do we need to add yet another one?

File: src/app/application/components/app-collection/AppCollection.vue [hunk 4]
```diff
@@ -71,6 +76,13 @@ const slots = useSlots()
 
 const items = ref<unknown[] | undefined>(props.items)
 const cacheKey = ref<number>(0)
+
+const remainingSlots = computed(() => {
+  const { toolbar, ...remainingSlots } = slots
+
+  return remainingSlots
+})
+
```
Comment (johncowen) [inline src/app/application/components/app-collection/AppCollection.vue:85]: If we can either do the `key === 'toolbar'`, or even better just reuse the `.k-table-toolbar` selector, when we can get rid of this code.

File: src/app/application/components/app-collection/AppCollection.vue [hunk 5]
```diff
@@ -84,11 +96,28 @@ const click = (e: MouseEvent) => {
...
   }
 }
+
+function getCellAttributes({ headerKey }: any) {
+  return {
+    class: {
+      [`${headerKey}-column`]: true,
+    },
+  }
+}
```
Comment (johncowen) [inline src/app/application/components/app-collection/AppCollection.vue:106]: I would rather keep the `class` inline please.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 3 above)
Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:33]: > Could we just check if key equals toolbar and if so add the wrapping div?

Technically, yes, but I don’t see how that’s better. Leaving this as is.

> In fact from what I can see the KTable toolbar already has a wrapping k-table-toolbar, do we need to add yet another one?

`.k-table-toolbar` has styles attached to it that I don’t want to re-use. Specifically that every element in toolbar should have `display: flex` is not desirable in my book. Having a separate element here makes the layout of toolbar items robust and independent from changes in Kongponents.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 4 above)
Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:85]: See https://github.com/kumahq/kuma-gui/pull/1203#discussion_r1267649374 for why I’m not using the existing `.k-table-toolbar` selector for these styles.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 5 above)
Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:106]: Why? Right now, we keep moving more and more logic into templates where it doesn’t belong. It’s already hard enough to parse what’s going on with a `KTable` that is fed a bunch of props, but adding inline functions to that makes it even worse.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 3 above)
Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:33]: > .k-table-toolbar has styles attached to it that I don’t want to re-use. Specifically that every element in toolbar should have display: flex is not desirable in my book.

I'm not seeing this could you post a screengrab or something showing me these?

Screenshot 2023-07-19 at 08 57 11

lmk

Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:33]: You have access to https://github.com/Kong/kongponents, right?

https://github.com/Kong/kongponents/blob/main/src/components/KTable/KTable.vue#L986-L992

File: src/app/application/components/app-collection/AppCollection.vue (hunk 5 above)
Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:106]: AppCollection.vue is 94 lines long. The template/html in there is 36 lines long, the vast majority of those 36 lines are just setting an attribute.

How hard is that to parse? Harder to parse than our listing components used to be?

File: src/app/application/components/app-collection/AppCollection.vue (hunk 4 above)
Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:85]: Just to note for history that we are continuing this conversation in the single place that @kleinfreund referred to.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 5 above)
Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:106]: Templates should have hardly any logic in them at all. I’m just putting this function where it belongs.

If there is nothing particularly wrong with this, I’m leaving this as is.

File: src/app/application/components/app-collection/AppCollection.vue (hunk 3 above)
Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:33]: Yes I do, but thanks for linking thats super helpful

File: src/app/application/components/app-collection/AppCollection.vue (hunk 5 above)
Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:106]: Where is the logic?

Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue:106]: There isn’t really a point arguing this. Tell if this is a blocker for you and I’ll move it back.

Comment (johncowen) [reply src/app/application/components/app-collection/AppCollection.vue:106]: This time around its a blocker for me.

The reasons you have given for doing this:

1. It's hard to follow. Its not.
2. It's logic and therefore shouldn't be in the template. There's no logic.
3. There's nothing wrong with it. There was nothing wrong with it in the first place and no reason to change (see points 1 and 2)

Happy to hear more opinion, but as of right now, this is a pointless conversation, I think we can agree on that?

Happy to revisit once we have this stream of work complete. The code we are talking about is in a centralized place and not spread out in multiple places in the application.

Comment (kleinfreund) [reply src/app/application/components/app-collection/AppCollection.vue]: It’s not useful to argue this. I’ve moved it back.

**Comment by johncowen** [comment]
I would really like to approve this and get it merged, I thought it might be helpful to make a suggestion before doing that in the form of a commit, mostly amends to the my own code:

https://github.com/johncowen/kuma-gui/commit/a31752ee9d237a8405b8757d60d2d2e421419993

Changes:

1. FetcherProps > FetcherParams. As they are params not props
2. CellAttrParams, so we know what we are getting here.  I think this is something you would prefer also? `:cell-attrs` remains inlined for the moment.
3. Use `v-if` to do the slot dance instead of `computed` plus destructuring.
4. Apparently, from looking at the link you provided `:deep` seems to be the Vue sanctioned way of what I had been using `:is()` for (crossing the slot), so I've also made that change.

I would really love it if you could work with me to make those ^ changes here also. It achieves what we want functionally and I would be super happy to then approve/merge this.

They are small things and I would much prefer be best to do them here if you could (I want to avoid yet another PR after this, ideally I would have done them in the original PR for AppCollection) and then leave this as is until we get the stream of work complete. At which point we can revisit this component if we find other improvements can be made.

Let me know.

**Review by johncowen — APPROVED** [review]
Awesome thankyou!
